### DISCUSSION TITLE
ramda and code maintainability

### DISCUSSION THREAD

**Discussion by jeetiss** [diegomura/react-pdf#1565] [General]
hello, i have question about maintenance of `@react-pdf`

Ramda is used in all parts of project, but it creates a lot of problems.
What is going on in this example?

```js
R.ifElse(
  R.isNil,
  R.identity,
  R.compose(
    R.join(''),
    R.juxt([
      R.compose(
        R.toUpper,
        R.head,
      ),
      R.tail,
    ]),
  ),
);
```
i spent a lot of time to understand what is going on in code samples like that.

- ramda is difficult to read
- ramda is difficult to change
- ramda is super slow

this code can be replaced with couple lines of plain js, and it works hundred times faster 
```js
(str) => {
  if (str == null) return str
  return str[0].toUpperCase() + str.slice(1)
}
```

I like the `@react-pdf` architecture and the changes in v2 that bring new horizons, but i think that ramda doesn't fit `@react-pdf` at all. code should be [optimized for change](https://overreacted.io/optimized-for-change/) and more friendly 

@diegomura what do you think about? Let discuss the possible ways for migration away from ramda

**Comment by diegomura** [comment]
@jeetiss 100% agree with you on this. I like what ramda enforces in some things (fn composability, for example), but I agree that the types of scenarios you just shared it makes no sense and should be changed. Happy to accept changes while we work on other PRs (I've been doing it a bit, I know you too), or PRs just to improve this

**Comment by maxaggedon** [comment]
As I was trying to get some hints on performance bottlenecks, I played around with node profiler and saw a lot of ramda references in the output. And then I found this : https://news.ycombinator.com/item?id=28583319
So it seems that ramda might have a big part in bad performances. So on top of code maintanability, I think it is another strong argument for trying to reduce its usage in this awesome library ;)

**Comment by diegomura** [comment]
Hi @maxaggedon ! I agree. I thought I removed it completely already though. Do we still have it?

**Comment by maxaggedon** [comment]
We are not using the latest version for now, but good to know that it is already removed. We will soon create new types of documents for which we will use the latest version :)
